src/db.c

```diff
@@ -1112,8 +1114,17 @@ void scanGenericCommand(client *c, robj *o, unsigned long cursor) {
         while ((ln = listNext(&li))) {
             sds key = listNodeValue(ln);
             initStaticStringObject(kobj, key);
-            if (expireIfNeeded(c->db, &kobj, 0)) {
-                listDelNode(keys, ln);
+            /* Filter an element if it isn't the type we want. */
+            if (typename) {
+                robj* typecheck = lookupKeyReadWithFlags(c->db, &kobj, LOOKUP_NOTOUCH|LOOKUP_NONOTIFY);
+                char* type = getObjectTypeName(typecheck);
+                if (strcasecmp((char*) typename, type)) {
+                    listDelNode(keys, ln);
+                }
+            } else {
```

Comment:
maybe we can add a `continue`, instead of `else`, so that when this code is removed we don't need to de-indent the call to `expireIfNeeded`